### PR TITLE
Implementation of Error Policies.

### PR DESCRIPTION
**Pull request by karknu**
Supersedes #939 and #902.

### CODE REVIEW

File: ouroboros-network/src/Ouroboros/Network/ErrorPolicy.hs [hunk 1]
```diff
@@ -0,0 +1,313 @@
...
+    show (ErrorPolicy (_err :: e err -> SuspendDecision DiffTime)) =
+           "ErrorPolicy ("
+        ++ tyConName (typeRepTyCon (typeRep (Proxy :: Proxy err)))
+        ++ ")"
+
+
+evalErrorPolicy :: forall et e.
+                   Exception e
+                => et e
+                -> ErrorPolicy et
```
Comment (mrBliss) [inline ouroboros-network/src/Ouroboros/Network/ErrorPolicy.hs]: Hmm, I would try to be consistent with naming the type parameters (`et`, `e`, `err`). In this function, you use `e` for the exception type and `et` for `ErrorPolicy`'s `e` parameter, whereas in `ErrorPolicy` you have `err` for the exception type and `e` as a wrapper.
I don't kind care much about `et` vs. `e` vs. `err`, I just want it to be consistent :slightly_smiling_face:.

This comment applies to the whole file (and more), not just to this function in particular.

File: ouroboros-network/src/Ouroboros/Network/ErrorPolicy.hs [hunk 2]
```diff
@@ -0,0 +1,313 @@
...
+import           Control.Monad.Class.MonadSTM
+import           Control.Monad.Class.MonadTime
+import           Control.Tracer
+
+import           Data.Semigroup.Action
+
+import           Ouroboros.Network.Subscription.PeerState
+
+data ErrorPolicy e where
+     ErrorPolicy :: forall e err.
```
Comment (mrBliss) [inline ouroboros-network/src/Ouroboros/Network/ErrorPolicy.hs]: `e` is a type parameter but also `forall`-quantified here, is that necessary?

File: ouroboros-network/src/Ouroboros/Network/ErrorPolicy.hs [hunk 3]
```diff
@@ -0,0 +1,313 @@
...
+                      Exception err
+                   => (e err -> SuspendDecision DiffTime)
```
Comment (mrBliss) [inline ouroboros-network/src/Ouroboros/Network/ErrorPolicy.hs]: Thinking out loud: is this `e` even necessary here? Now when I have to write my `ErrorPolicy`s, I have to deal with this `e` each time, in my case it will always be `ApplicationException`. *Can't this `e` be a phantom parameter?*

File: ouroboros-network/src/Ouroboros/Network/ErrorPolicy.hs [hunk 4]
```diff
@@ -0,0 +1,313 @@
...
+
+-- | List of error policies for exception handling and a policy for handing
+-- application return values.
+--
+data ErrorPolicies m addr a = ErrorPolicies {
+    -- | Application Error Policies
+    epAppErrorPolicies  :: [ErrorPolicy ApplicationException]
+    -- | `connect` Error Policies
+  , epConErrorPolicies  :: [ErrorPolicy ConnectionException]
+  , epReturnCallback :: Time m -> addr -> a -> SuspendDecision DiffTime
```
Comment (mrBliss) [inline ouroboros-network/src/Ouroboros/Network/ErrorPolicy.hs]: See https://github.com/input-output-hk/ouroboros-network/pull/939/files?file-filters%5B%5D=.hs#r331943235
If this is part of the public API, please document it.

File: ouroboros-network/src/Ouroboros/Network/ErrorPolicy.hs (hunk 2 above)
Comment (mrBliss) [reply ouroboros-network/src/Ouroboros/Network/ErrorPolicy.hs]: If that somehow is necessary, please put `err` first, because I'd like to use type application to instantiate `err`.

File: ouroboros-network/src/Ouroboros/Network/NodeToNode.hs [hunk 5]
```diff
@@ -287,24 +317,29 @@ ipSubscriptionWorker
           peerid
           NodeToNodeProtocols
           IO BL.ByteString x y)
-    -> IO void
+    -> IO ()
```
Comment (mrBliss) [inline ouroboros-network/src/Ouroboros/Network/NodeToNode.hs]: What's the reason for this change? (throughout this file)

Comment (karknu) [reply ouroboros-network/src/Ouroboros/Network/NodeToNode.hs]: #902 Was written with ().

File: ouroboros-network/src/Ouroboros/Network/ErrorPolicy.hs (hunk 2 above)
Comment (karknu) [reply ouroboros-network/src/Ouroboros/Network/ErrorPolicy.hs]: Yes, `forall` for `e` is needed, but I will put `err` first.

File: ouroboros-network/src/Ouroboros/Network/ErrorPolicy.hs (hunk 3 above)
Comment (karknu) [reply ouroboros-network/src/Ouroboros/Network/ErrorPolicy.hs]: I don't know if that is possible.

File: ouroboros-network/src/Ouroboros/Network/ErrorPolicy.hs (hunk 1 above)
Comment (karknu) [reply ouroboros-network/src/Ouroboros/Network/ErrorPolicy.hs]: I will change to use `et` for exception type and e for exception everywhere.

File: ouroboros-network/src/Ouroboros/Network/ErrorPolicy.hs (hunk 4 above)
Comment (karknu) [reply ouroboros-network/src/Ouroboros/Network/ErrorPolicy.hs]: @coot ^ Not sure what your intention is here.

File: ouroboros-network/src/Ouroboros/Network/ErrorPolicy.hs (hunk 2 above)
Comment (mrBliss) [reply ouroboros-network/src/Ouroboros/Network/ErrorPolicy.hs]: I see. Ah, it's because of the `forall`, just like for functions, when there's a rank-1 `forall`, all type variables need to be explicitly quantified.

I have just checked locally that the `forall` isn't even needed. It even expects `err` as  its first type argument! I prefer the explicit `forall` here, though, to signal that there is an explicit quantification.

File: ouroboros-network/src/Ouroboros/Network/NodeToNode.hs (hunk 5 above)
Comment (mrBliss) [reply ouroboros-network/src/Ouroboros/Network/NodeToNode.hs]: ok. @coot ^

File: ouroboros-network/src/Ouroboros/Network/ErrorPolicy.hs (hunk 4 above)
Comment (coot) [reply ouroboros-network/src/Ouroboros/Network/ErrorPolicy.hs]: I think we can either remove it or pass a a dummy decision (delay by `0sec`); The initial intention was to do something when a mini-protocol finishes in a legit way (sending one of the `Done` messages), but this never happens the consensus always throws exceptions (is that right @mrBliss - I'd like you to double check this).

Comment (coot) [reply ouroboros-network/src/Ouroboros/Network/ErrorPolicy.hs]: I checked `ChainSync`, `TxSubsmission` and `BlockFetch` protocols: their clients which are used in `ouroboros-consensus`  never return, so the question is just remains for the clients. But there if one terminates a protocol, I think it's ok to just terminate the whole application, so no action is needed. Let's just remove `epReturnCallback`.

Comment (coot) [reply ouroboros-network/src/Ouroboros/Network/ErrorPolicy.hs]: @karknu I would not package both `epAppErrorPolicies` and `epConErrorPolicies` into one record. The `NodeToNode` should only expose `apAppErrorPolicies` which will be filled by `ouroboros-consensus` (as they have all the exceptions for that) and we'll supply `apConErrorPolicy` (which will contain some policy for `IOException`s).

File: ouroboros-network/src/Ouroboros/Network/NodeToNode.hs (hunk 5 above)
Comment (coot) [reply ouroboros-network/src/Ouroboros/Network/NodeToNode.hs]: I'd prefer a `void` or `Void`

File: ouroboros-network/src/Ouroboros/Network/ErrorPolicy.hs (hunk 3 above)
Comment (coot) [reply ouroboros-network/src/Ouroboros/Network/ErrorPolicy.hs]: I think it can be done with just a phantom parameter. @karknu but if you'll find your self in a corner then we can clean it afterwards.

File: ouroboros-network/src/Ouroboros/Network/ErrorPolicy.hs (hunk 4 above)
Comment (mrBliss) [reply ouroboros-network/src/Ouroboros/Network/ErrorPolicy.hs]: @coot The `ChainSyncClient` that we implemented never finishes and has return type `... Void`. We use the stock `BlockFetch` and `TxSubmission` clients from the network layer, so they should also never finish. I have just verified this by trying to changing their return type to `forall void. ... void` too.

Let me make a separate PR for that change.

Comment (karknu) [reply ouroboros-network/src/Ouroboros/Network/ErrorPolicy.hs]: @coot Re. exposing `epConErrorPolicies`. I resurrected this PR in order to support error handling over NodeToClient for explorer. The suspension timeouts should vary depending on the application and network transport technology. A connection error to your local node is not the same as a connection error to a public relay.

Comment (coot) [reply ouroboros-network/src/Ouroboros/Network/ErrorPolicy.hs]: @karknu ok I am convinced;
@mrBliss I am glad to see that PR.

File: ouroboros-network/src/Ouroboros/Network/ErrorPolicy.hs (hunk 3 above)
Comment (coot) [reply ouroboros-network/src/Ouroboros/Network/ErrorPolicy.hs]: I addressed this in the last commit.

File: ouroboros-network/src/Ouroboros/Network/NodeToNode.hs (hunk 5 above)
Comment (coot) [reply ouroboros-network/src/Ouroboros/Network/NodeToNode.hs]: Addressed

File: ouroboros-consensus/src/Ouroboros/Consensus/Node.hs [hunk 6]
```diff
@@ -295,8 +301,11 @@ initNetwork registry nodeArgs kernel RunNetworkArgs{..} = do
 
     -- serve downstream nodes
     connTable  <- newConnectionTable
+    peerStatesVar <- newPeerStatesVar
+    -- every 200s clean peer states
+    forkLinkedThread registry (NodeToNode.cleanPeerStates 200 peerStatesVar)
```
Comment (mrBliss) [inline ouroboros-consensus/src/Ouroboros/Consensus/Node.hs]: Don't ignore the returned thread (a warning currently, causing CI to fail) but add it to the list of `threads` at the end of the function.

File: ouroboros-consensus/src/Ouroboros/Consensus/Node.hs [hunk 7]
```diff
@@ -329,42 +338,56 @@ initNetwork registry nodeArgs kernel RunNetworkArgs{..} = do
...
         rnaMuxLocalTracer
         rnaHandshakeLocalTracer
+        rnaErrorPolicyTracer
         connTable
+        peerStatesVar
         rnaMyLocalAddr
         rnaMkPeer
         nodeToClientVersionData
         (localResponderNetworkApplication networkApps)
+        nullErrorPolicies
```
Comment (mrBliss) [inline ouroboros-consensus/src/Ouroboros/Consensus/Node.hs]: Is `ouroboros-network` usable with this policy? Will this break things for consumers of this library when this PR is merged without putting the correct policy in place?

File: ouroboros-consensus/src/Ouroboros/Consensus/Node.hs [hunk 8]
```diff
@@ -329,42 +338,56 @@ initNetwork registry nodeArgs kernel RunNetworkArgs{..} = do
...
         rnaMuxTracer
         rnaHandshakeTracer
+        rnaErrorPolicyTracer
         connTable
+        peerStatesVar
         myAddr
         rnaMkPeer
         nodeToNodeVersionData
         (responderNetworkApplication networkApps)
+        nullErrorPolicies
```
Comment (mrBliss) [inline ouroboros-consensus/src/Ouroboros/Consensus/Node.hs]: Same question as for `runLocalServer`

**Comment by coot** [comment]
> @coot can you help me fill in the holes (_a.._e)?
All are `DiffTime`.

File: ouroboros-network/src/Ouroboros/Network/Subscription/PeerState.hs [hunk 9]
```diff
@@ -253,6 +256,69 @@ instance Eq addr
...
+                   , MonadTime  m
+                   , MonadTimer m
+                   )
+                => DiffTime
+                -> StrictTVar m (PeerStates m addr)
+                -> m ()
+cleanPeerStates interval v = go
+  where
+    go = do
+      threadDelay interval
```
Comment (dcoutts) [inline ouroboros-network/src/Ouroboros/Network/Subscription/PeerState.hs]: Do we have to do this with polling? Can't we do use STM to wait for the state to change?

File: ouroboros-network/src/Ouroboros/Network/Server/Socket.hs [hunk 10]
```diff
@@ -77,77 +88,93 @@ type Main st t = st -> STM t
...
 -- the server can stop pulling from the queue, without causing the child
 -- threads to hang attempting to write to it.
-type ResultQ r = STM.TQueue (Result r)
+type ResultQ addr r = STM.TQueue (Result addr r)
 
 -- | The product of a spawned thread. We catch all (even async) exceptions.
-data Result r = Result
+data Result addr r = Result
   { resultThread :: !(Async ())
+  , resultAddr   :: !addr
```
[the remark this answers is not in the document]
Comment (coot) [reply ouroboros-network/src/Ouroboros/Network/Server/Socket.hs]: We need to know the address from which the result is coming from.  It is used to form `ApplicationError` [here](https://github.com/input-output-hk/ouroboros-network/pull/1127/files#diff-37c626674753d27fd06d21f7c2439ec6R467) and passed to [completeApplicationTx](https://github.com/input-output-hk/ouroboros-network/pull/1127/files#diff-7dc3e056b4d6313a7eb3336911e4a13aR162).  This way the error policy will apply to the given address (we use `IP` addresses as the node identification).

File: ouroboros-network/src/Ouroboros/Network/Server/Socket.hs [hunk 11]
```diff
@@ -56,11 +60,18 @@ data HandleConnection channel st r where
...
+--
+-- It is needed only because 'BeginConnection' does not have access to the
+-- thread which runs the application.
+--
+type ApplicationStart addr st = addr -> Async () -> st -> STM st
 
 -- | How to update state when a connection finishes. Can use `throwSTM` to
 -- terminate the server.
-type CompleteConnection st r = Either SomeException r -> st -> STM st
+type CompleteConnection addr st r = Result addr r -> st -> STM (st, IO ())
```
[the remark this answers is not in the document]
Comment (coot) [reply ouroboros-network/src/Ouroboros/Network/Server/Socket.hs]: This is used for two things:
* tracing
* cancelling threads as a result of evaluation of an error policy

It might be cleaner to just return trace points and list of threads, and do that more explicitely.

File: ouroboros-network/src/Ouroboros/Network/Server/Socket.hs [hunk 12]
```diff
@@ -192,25 +219,26 @@ mainLoop resQ threadsVar statusVar complete main =
   connectionTx = do
     result <- STM.readTQueue resQ
     st <- STM.readTVar statusVar
-    !st' <- complete (resultValue result) st
+    (!st', io) <- complete result st
     STM.writeTVar statusVar st'
     -- It was inserted by `spawnOne`.
     STM.modifyTVar' threadsVar (Set.delete (resultThread result))
-    pure $ mainLoop resQ threadsVar statusVar complete main
+    pure $ io >> mainLoop resQ threadsVar statusVar complete main
```
[the remark this answers is not in the document]
Comment (coot) [reply ouroboros-network/src/Ouroboros/Network/Server/Socket.hs]: it's only `traceWith ... >> traverse_ cancel threadIds`, I'll make it more explicit.

File: ouroboros-network/src/Ouroboros/Network/Subscription/PeerState.hs (hunk 9 above)
Comment (coot) [reply ouroboros-network/src/Ouroboros/Network/Subscription/PeerState.hs]: We could, but we'll need to use `unsafeIOToSTM`.  We need `IO` to read the current time with `getMonotonicTime`.  We use an unsafe foreign call `getMonotonicNSec` so I think it's safe to wrap it in `unsafeIOToSTM`:
```
forever $ atomically $ do
  s <- readTVar v
  t <- unsafeIOToSTM getMonotonicTime
  ...
```
